controllers/vm_migration_controller.go

```diff
@@ -0,0 +1,193 @@
+/*
+Copyright 2024.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strings"
+
+	networkv1alpha1 "github.com/NCCloud/tabby-cni/api/v1alpha1"
+	"github.com/j-keck/arping"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	virtv1 "kubevirt.io/api/core/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+// VirtualMachineReconciler makes sure the VM has a correct mac address on the switch
+type VirtualMachineReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=kubevirt.io,resources=virtualmachineinstances,verbs=get;list;watch
+
+func (r *VirtualMachineReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = log.FromContext(ctx)
+
+	log.Log.Info(fmt.Sprintf("VirtualMachine: Reconcile Virtual Machine %+v", req))
+
+	virtualMachineInstance := &virtv1.VirtualMachineInstance{}
+	err := r.Get(ctx, req.NamespacedName, virtualMachineInstance)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	virtualMachineNetworks := virtualMachineInstance.Spec.Networks
+
+	for i := range virtualMachineNetworks {
+		multusNetwork := virtualMachineNetworks[i].NetworkSource.Multus
+		// network type can be Multus or Pod. We will work only with Multus
+		if multusNetwork == nil {
+			continue
+		}
+
+		namespacedNetworkName := getNamespacedNetworkName(multusNetwork.NetworkName)
+
+		network := &networkv1alpha1.Network{}
+		err := r.Get(ctx, namespacedNetworkName, network)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		networkIpMasq := network.Spec.IpMasq
+		// send a garp request only if IP masquerading is enabled
+		if !networkIpMasq.Enabled {
+			continue
+		}
+
+		// arping -A -i <interface-name> -S 169.254.1.1 169.254.1.1
+		interfaceName := networkIpMasq.Bridge
+		log.Log.Info(
```

Comment:
Variadic log statement can be used

Reply:
strange, in variadic notation you don't need place holders, but anyways just ignore this I will try this separately :)